cli/src/commands/metadata/retain.rs

```diff
@@ -0,0 +1,603 @@
+
+    // Strip whichever variants we can get away with:
+    for id in outer_enum_ids.iter() {
+        strip_variants_in_enum_type(
+            &mut metadata.get_portable_registry_mut(), 
+            &keep_pallet, 
+            id
+        )
+    }
+
+    // Iterate over the type IDs again (no need to deeply recurse this time), and retain
+    // any of these 9and their children) in the type registry:
+    let keep_these_ids: BTreeSet<u32> = metadata.iterate_all_type_ids().collect();
+    metadata.get_portable_registry_mut().retain(|id | keep_these_ids.contains(&id));
+}
+
+/// This trait is implemented for metadatas, and its purpose is to hand back iterators over
+/// all of the type IDs (doesn't need to recurse into them) that are used in the metadata,
+/// so that we know which ones we need to keep around in the type registry (and thus which 
+/// ones we can remove).
+/// 
+/// There are two methods that need implementing:
+/// - [`IterateTypeIds::iterate_type_ids_to_consider_for_stripping`] should return all type IDs
+///   that, if they contain outer enum types, would mean that we cannot strip varaints out of
+///   said outer enum types.
+/// - [`IterateTypeIds::iterate_type_ids_ignored_for_stripping`] should return all of the type
+///   IDs found in the metadata but _not_ returned from the above call.
+/// 
+/// Which types do we need to consider when it comes ot knowing whether to strip our outer
+/// enum variants (ie remove the variants representing pallets that we are not interested in)?
+/// Basically, any type that the user could obtain through Subxt which may contains an outer enum, 
+/// and which Subxt needs to decode as a result of this. (If we strip things from the outer enums, then
+/// they won't necessarily be able to decode everything we get given back any more!). Examples:
+/// - Runtime API output types (but not input types: we can still encode them ok regardless)
+/// - Storage keys and values (both are decoded)
+/// 
+/// One exception here is that Subxt can decode extrinsics in blocks, but if metadata is stripped
+/// then some extrinsics won't be decodable any more, and that's fine. This may also happen if the 
+/// metadata being used is behind the latest runtime version, and in any case, users should not
+/// strip metadata if they want to decode extrinsics freom _any_ pallet.
+/// 
+/// When implementing this, if you're ever not sure whether a type ID should be considered for
+/// stripping or not, always assume that it is to play it safe. If _everything_ is in 
+/// [`IterateTypeIds::iterate_type_ids_to_consider_for_stripping`], then things will still work
+/// (we'll just end up stripping fewer things from the metadata).
+trait IterateTypeIds {
+    /// This should return an iterator over all type IDs which, if seen, should prevent us
+    /// from stripping variants from RuntimeCall/Error/Event enums ("outer enums").
+    fn iterate_type_ids_to_consider_for_stripping(&self) -> impl Iterator<Item=u32>;
```

Comment:
I don't think you need to repeat the type ids for every method here

I think `iter_for_stripping`, `iter_not_stripping` and `iter_all/iter` would be sufficient but your call :)

Reply:
Chanegd to `iter_type_ids_mut` (I realised it needed to be mutable for updating the IDs after)